File: src/hidden_shuffle.f90
!> Based on the paper
!! https://wrap.warwick.ac.uk/id/eprint/150064/1/WRAP-sequential-random-sampling-revisited-hidden-shuffle-method-2021.pdf
!! TODO:
!! * Make some notes on the theory, such that the basic premise is understood
!! * Plot the number distribution and confirm it is normal

module hidden_shuffle_m
    use iso_fortran_env
    implicit none
    private
    public :: hidden_shuffle

contains

    subroutine hidden_shuffle(m_in, N, result, seed)
        integer, intent(in) :: m_in !< Number of samples to take
        integer, intent(in) :: N !< Range to sample from [0,N)
        integer, intent(out) :: result(:)
        integer, intent(in), optional :: seed(:)

        integer :: H, L, i, s_old, s, cnt, m
        real(real64) :: N_minus_m, q, p_i, random_num, a, F

        if (present(seed)) then
            call random_seed(put=seed)
        else
            call random_seed()
        endif

        m = m_in
        cnt = 0
        H = 0
        i = 0
        N_minus_m = real(N - m, real64)

        ! Step 1 (note what this does)
        if (N > m) then
            H = m
            do while (i < m)
                ! Generate random number between 0 and 1
                call random_number(random_num)

                q = 1._real64 - N_minus_m / real(N - i, real64)
                i = i + int(log_b(random_num, real(1 - q, real64)))
                p_i = 1._real64 - N_minus_m / real(N - i, real64)

                call random_number(random_num)
                if ((i < m) .and. (random_num < p_i / q)) then
                    H = H - 1
                end if
                i = i + 1
            end do
        end if

        ! Step 2. Draw high items. Basically looks like reservoir sampling
        L = m - H
        a = 1._real64
        do while(H > 0)
            s_old = m + int(a * N_minus_m)
            call random_number(random_num)
            a = a * random_num**(1._real64 / real(H, real64))
            s = m + int(a * N_minus_m)
            if (s < s_old) then
                cnt = cnt + 1
                result(cnt) = N - 1 - S
                if (cnt == m) return
            else
                L = L + 1
            end if
            H = H - 1
        end do

        ! Step 3. Draw low items
        do while(L > 0)
            call random_number(random_num)
            s = 0
            f = real(L, real64) / real(m, real64)
            do while(f < random_num .and. s < (m - L - 1))
                f = 1._real64 - (1._real64 - real(L, real64) / real(m - s, real64)) * (1._real64 - f)
                s = s + 1
            end do
            L = L - 1
            m = m - s -1
            cnt = cnt + 1
            result(cnt) = N - 1 - m
            if (cnt == m) return
        end do

    end subroutine hidden_shuffle

    !> Log of a specified base
    function log_b(x, b) result(l)
        real(real64) :: x, b
        real(real64) :: l

        l = log(x) / log(b)

    end function log_b

end module hidden_shuffle_m

program test
    use hidden_shuffle_m
    integer :: i, m
    integer, allocatable :: result(:)
    ! To fix the seed
    !integer :: min_seed_size = 10
    !call random_seed(put=min_seed_size)
    !allocate(seed(min_seed_size), source=[(i, i=1, min_seed_size)])

    m = 100
    allocate(result(m))
    call hidden_shuffle(m, 10000, result)

    ! Transform to limit [i, j) by setting N = j - i
    ! For [1, np], this means passing N = np
    ! hen adding i = 1 to the return values
    do i = 1, m
        write(*, *) result(i) + 1
    end do

end program test

File: src/python/hidden_shuffle.py
from math import log
import random
from random import uniform


def hidden_shuffle(n, N):
    """Random sampling with replacement

     Obtain a sample of size n, from population size N
    Values range [0, N).

    See https://www.pcg-random.org/posts/bounded-rands.html
    for how to simply map [0, N). to [i, j]

    :param n:
    :param N:
    :return:
    """
    H = 0
    i = 0
    N_minus_m = float(N - n)
    if N > n:
        H = n
        while i < n:
            q = 1.0 - N_minus_m / float(N-i)
            i = i + int(log(uniform(0, 1), 1-q) )
            p_i = 1.0 - N_minus_m / float(N-i)
            if i < n and uniform(0, 1) < p_i / q:
                H = H-1
            i = i + 1
    L = n - H; a = 1.0
    while H > 0:  # STEP 2: draw high−items
        S_old = n + int(a * N_minus_m)
        a = a * uniform(0, 1)**(1.0 / H)
        S = n + int(a * N_minus_m)
        if S < S_old:
            yield N - 1 - S
        else:
            L = L + 1  # duplicate detected
        H = H-1
    while L > 0:  # STEP 3: draw low−items
        u = uniform(0, 1)
        s = 0
        F = float(L) / float(n)
        while F < u and s < (n-L-1):
            F = 1 - (1 - float(L) / float(n-s)) * (1 - F)
            s = s + 1
        L = L-1
        n = n-s-1
        yield (N-1) - n


random.seed(42)  # You can use any number as a seed

ref = [ 148, 179, 218, 229, 475, 561, 907, 1054, 1121,
        1468, 1619, 1660, 1717, 1860, 1909, 1929, 2399,
        2418, 2452, 2551, 2722, 2726, 2827, 3045, 3255,
        3270, 3316, 3335, 3364, 3422, 3425, 3517, 3573,
        3591, 3637, 3652, 3706, 3741, 4045, 4187, 4306,
        4544, 4680, 4889, 5004, 5045, 5137, 5227, 5368,
        5486, 5492, 5532, 5577, 5740, 5769, 5936, 6024,
        6025, 6067, 6123, 6159, 6176, 6201, 6350, 6682,
        6789, 6911, 7053, 7058, 7071, 7181, 7222, 7312,
        7321, 7400, 7535, 7675, 7732, 7866, 7920, 7931,
        8029, 8189, 8264, 8521, 8810, 8993, 9031, 9051,
        9128, 9353, 9423, 9424, 9481, 9484, 9504, 9857,
        9869, 9943, 9974]

result = [value for value in hidden_shuffle(100, 10000)]

assert result == ref
